libraries/freertos_plus/standard/freertos_plus_tcp/source/portable/NetworkInterface/STM32Fxx/NetworkInterface.c

```diff
@@ -701,32 +692,54 @@ const TickType_t xBlockTimeTicks = pdMS_TO_TICKS( 50u );
 				{
 					/* Copy the bytes. */
 					memcpy( ( void * ) pxDmaTxDesc->Buffer1Addr, pxDescriptor->pucEthernetBuffer, ulTransmitSize );
-					pxDmaTxDesc->Status |= ETH_DMATXDESC_CIC_TCPUDPICMP_FULL;
 				}
 				#else
 				{
+					configASSERT( bReleaseAfterSend != 0 );
+
 					/* Move the buffer. */
 					pxDmaTxDesc->Buffer1Addr = ( uint32_t )pxDescriptor->pucEthernetBuffer;
-					/* Ask to set the IPv4 checksum.
-					Also need an Interrupt on Completion so that 'vClearTXBuffers()' will be called.. */
-					pxDmaTxDesc->Status |= ETH_DMATXDESC_CIC_TCPUDPICMP_FULL | ETH_DMATXDESC_IC;
 					/* The Network Buffer has been passed to DMA, no need to release it. */
 					bReleaseAfterSend = pdFALSE_UNSIGNED;
 				}
 				#endif /* ipconfigZERO_COPY_TX_DRIVER */
 
+				/* Ask to set the IPv4 checksum.
+				Also need an Interrupt on Completion so that 'vClearTXBuffers()' will be called.. */
+				#if( ipconfigDRIVER_INCLUDED_TX_IP_CHECKSUM != 0 )
+				{
+					pxDmaTxDesc->Status |= ETH_DMATXDESC_CIC_TCPUDPICMP_FULL | ETH_DMATXDESC_IC;
+				}
+				#else
+				{
+					pxDmaTxDesc->Status &= ~( ( uint32_t ) ETH_DMATXDESC_CIC );
+					pxDmaTxDesc->Status |= ETH_DMATXDESC_IC;
+				}
+				#endif
+				
+
 				/* Prepare transmit descriptors to give to DMA. */
 
 				/* Set LAST and FIRST segment */
 				pxDmaTxDesc->Status |= ETH_DMATXDESC_FS | ETH_DMATXDESC_LS;
 				/* Set frame size */
 				pxDmaTxDesc->ControlBufferSize = ( ulTransmitSize & ETH_DMATXDESC_TBS1 );
+
+				#if( NETWORK_BUFFERS_CACHED	!= 0 )
+				{
+				BaseType_t xlength = CACHE_LINE_SIZE * ( ( ulTransmitSize + NETWORK_BUFFER_HEADER_SIZE + CACHE_LINE_SIZE - 1 ) / CACHE_LINE_SIZE );
+				uint32_t *pulBuffer = ( uint32_t )( pxDescriptor->pucEthernetBuffer - NETWORK_BUFFER_HEADER_SIZE );
+					cache_clean_invalidate_by_addr( pulBuffer, xlength );
+				}
+				#endif
+
 				/* Set Own bit of the Tx descriptor Status: gives the buffer back to ETHERNET DMA */
 				pxDmaTxDesc->Status |= ETH_DMATXDESC_OWN;
 
 				/* Point to next descriptor */
 				xETH.TxDesc = ( ETH_DMADescTypeDef * ) ( xETH.TxDesc->Buffer2NextDescAddr );
-	
+				/* Ensure completion of memory access */
+				__DSB();
```

Comment:
I'd consider creating a macro for handling these DMA descriptors to reduce the probability of a future developer mishandling them (since this can be difficult to debug/find), maybe something like:

```c
#define ETH_DMA_DESCRIPTOR_ACCESS(operation) \
do {\
   (operation); \
  __DSB(); \
} while( 0 );
```

...and then:

```c
ETH_DMA_DESCRIPTOR_ACCESS(pxDmaTxDesc->Status |= ETH_DMATXDESC_OWN);
```

Reply:
I would like to postpone this for a while. There are also readback's that have the same function. But you're right, it is good to protect these construction by putting them a `#define`